### ISSUE TITLE
Inefficient Or handling when one clause is None

### ISSUE THREAD

**Issue by stevebarham** [npgall/cqengine#167]
(edit - rewrote for clarity after previously dropping out multiple stream-of-consciousness comments)

This might seem like a bit of an edge case, but it bit me today, so thought I'd write it up. 

The block of code starting at https://github.com/npgall/cqengine/blob/master/code/src/main/java/com/googlecode/cqengine/engine/CollectionQueryEngine.java#L1009 executes a collection scan, when the cost of an Or query is Integer.MAX_VALUE. 

The None query is only handled by the FallbackIndex. This has an efficient implementation for None in FallbackIndex#retrieve, but still has Integer.MAX_VALUE cost. 

This means that where we have an Or query of the following shape:

Or(None, Q)
Or(Q, None)

We fall back to a collection scan, irrespective of whether Q could be efficiently retrieved via an index. 

This shape of query can be easily constructed when an empty collection is passed to an In query. I hit this today with a query resembling this:

```java
Collection<Integer> delete = entitiesToExplicitlyDelete;
Collection<Integer> update = entitiesToUpsert.stream.map(e -> e.id).collect(toSet());
collection.update(collection.retrieve(or(in(key, delete), in(key, update))), emptySet());
```

Where entitiesToExplicitlyDelete is empty, we do a collection scan.

I can guard against this in my calling code, but it seems like we should be able to optimise this in cqengine. My proposal is to:

- Implement an EmptyIndex, which only supports None queries, and which returns an empty result set with zero retrieval and merge cost
- Return this index alongside the fallback index in CollectionQueryEngine#getIndexesOnAttribute

Let me know if this sounds sane, and I'll put a PR together.

**Comment by stevebarham** [comment]
p.s. thanks for cqengine - it's fantastic.

**Comment by stevebarham** [comment]
On closer inspection, this should be handled already by the FallbackIndex's retrieve method, which has a specific case for None. Let me get more detail on the issue.

**Comment by stevebarham** [comment]
Got it - this looks like an edge case where we have queries of the shape:

or(none, something)

This is caused by the block starting at https://github.com/npgall/cqengine/blob/master/code/src/main/java/com/googlecode/cqengine/engine/CollectionQueryEngine.java#L1009

At this point in our query, the FallbackIndex has Integer.MAX_VALUE retrieval cost, so the engine knows here that at least one branch of the Or evaluation will require a collection scan. 

Implementing an EmptyIndex per my initial comment would dodge this block of code, as the LHS of the query would yield a retrieval cost of 0, rather than Integer.MAX_VALUE.

**Comment by npgall** [comment]
Hi Steve,

That is a really great suggestion, and you are spot on with your findings!

I think we can just fix this in FallbackIndex itself. 

Normally (by definition) the `FallbackIndex` does not have a lot of information to work with when it's asked to compute the merge cost (an estimate of how many objects match the query), so it returns `Integer.MAX_VALUE`.

However as you suggested, None is actually a special case where we know how many objects it will match. So it would be a safe, one-liner optimization, to have FallbackIndex return merge cost 0 for None queries.

I'll make that change and cut a new release. Thanks for this great suggestion!

**Comment by npgall** [comment]
Done. This is now released in CQEngine 2.12.4. Thanks again!

**Comment by stevebarham** [comment]
Nice one - thank you!
